firebase-app-distribution/src/main/java/com/google/firebase/app/distribution/FirebaseAppDistribution.java

```diff
@@ -294,113 +396,39 @@ AppDistributionReleaseInternal getCachedNewRelease() {
-          synchronized (updateIfNewReleaseTaskLock) {
-            // show download progress in notification manager
-            updateApp(true)
-                .addOnProgressListener(this::postProgressToCachedUpdateIfNewReleaseTask)
-                .addOnSuccessListener(unused -> setCachedUpdateIfNewReleaseResult())
-                .addOnFailureListener(cachedUpdateIfNewReleaseTask::setException);
-          }
-        });
-
-    updateDialog.setButton(
-        AlertDialog.BUTTON_NEGATIVE,
-        context.getString(R.string.update_no_button),
-        (dialogInterface, i) -> dismissUpdateDialogCallback());
-
-    updateDialog.setOnCancelListener(
-        dialogInterface -> {
-          dismissUpdateDialogCallback();
-        });
-
-    updateDialog.show();
-    updateDialogShown = true;
-    synchronized (updateIfNewReleaseTaskLock) {
-      return cachedUpdateIfNewReleaseTask;
-    }
-  }
-
-  private void dismissUpdateDialogCallback() {
-    synchronized (updateIfNewReleaseTaskLock) {
-      postProgressToCachedUpdateIfNewReleaseTask(
-          UpdateProgress.builder()
-              .setApkFileTotalBytes(UNKNOWN_RELEASE_FILE_SIZE)
-              .setApkBytesDownloaded(UNKNOWN_RELEASE_FILE_SIZE)
-              .setUpdateStatus(UpdateStatus.UPDATE_CANCELED)
-              .build());
-      setCachedUpdateIfNewReleaseCompletionError(
-          new FirebaseAppDistributionException(
-              ErrorMessages.UPDATE_CANCELED, Status.INSTALLATION_CANCELED));
-    }
-  }
-
-  private void setCachedUpdateIfNewReleaseCompletionError(FirebaseAppDistributionException e) {
+  private void setCachedUpdateIfNewReleaseCompletionError(Exception e) {
     synchronized (updateIfNewReleaseTaskLock) {
       safeSetTaskException(cachedUpdateIfNewReleaseTask, e);
     }
-    dismissUpdateDialog();
-  }
-
-  private void setCachedUpdateIfNewReleaseCompletionError(
-      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
-    if (e instanceof FirebaseAppDistributionException) {
-      setCachedUpdateIfNewReleaseCompletionError((FirebaseAppDistributionException) e);
-    } else {
-      setCachedUpdateIfNewReleaseCompletionError(defaultFirebaseException);
-    }
+    dismissDialogs();
   }
 
   private void postProgressToCachedUpdateIfNewReleaseTask(UpdateProgress progress) {
     synchronized (updateIfNewReleaseTaskLock) {
-      cachedUpdateIfNewReleaseTask.updateProgress(progress);
+      if (cachedUpdateIfNewReleaseTask != null && !cachedUpdateIfNewReleaseTask.isComplete()) {
+        cachedUpdateIfNewReleaseTask.updateProgress(progress);
+      }
     }
   }
 
   private void setCachedUpdateIfNewReleaseResult() {
     synchronized (updateIfNewReleaseTaskLock) {
       safeSetTaskResult(cachedUpdateIfNewReleaseTask);
     }
-    dismissUpdateDialog();
+    dismissDialogs();
   }
 
-  private void dismissUpdateDialog() {
+  private void dismissDialogs() {
+    if (signInDialog != null && signInDialog.isShowing()) {
```

Comment:
Addition to my comment above about dialogs. Is there a reason why theres some difference in logic for checking to dismiss the dialogs?

Reply:
Good catch! I don't think so... I'm not sure why we are using a custom boolean flag `updateDialogShown` instead of relying on the built-in `updateDialog.isShowing()` bool.